src/arithmetic/algebraic_expression/algebraic_expression_eval.c

```diff
@@ -0,0 +1,255 @@
+	bool res_in_use = false;        // Can we use `res` for intermidate evaluation.
+
+	GrB_Descriptor_new(&desc);
+
+	// Get left and right operands.
+	AlgebraicExpression *left = CHILD_AT(exp, 0);
+	AlgebraicExpression *right = CHILD_AT(exp, 1);
+
+	/* If left operand is a matrix, simply get it.
+	 * Otherwise evaluate left hand side using `res` to store LHS value. */
+	if(left->type == AL_OPERAND) {
+		a = left->operand.matrix;
+	} else {
+		if(left->operation.op == AL_EXP_TRANSPOSE) {
+			a = left->operation.children[0]->operand.matrix;
+			GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+		} else {
+			a = _AlgebraicExpression_Eval(left, res);
+			res_in_use = true;
+		}
+	}
+
+	/* If right operand is a matrix, simply get it.
+	 * Otherwise evaluate right hand side using `res` if free or create an additional matrix to store RHS value. */
+	if(right->type == AL_OPERAND) {
+		b = right->operand.matrix;
+	} else {
+		if(right->operation.op == AL_EXP_TRANSPOSE) {
+			b = right->operation.children[0]->operand.matrix;
+			GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+		} else if(res_in_use) {
+			// `res` is in use, create an additional matrix.
+			GrB_Matrix_nrows(&nrows, a);
+			GrB_Matrix_ncols(&ncols, a);
+			info = GrB_Matrix_new(&inter, GrB_BOOL, nrows, ncols);
+			if(info != GrB_SUCCESS) {
+				fprintf(stderr, "%s", GrB_error());
+				assert(false);
+			}
+			b = _AlgebraicExpression_Eval(right, inter);
+		} else {
+			// `res` is not used just yet, use it for RHS evaluation.
+			b = _AlgebraicExpression_Eval(right, res);
+		}
+	}
+
+	// Perform addition.
+	if(GrB_eWiseAdd_Matrix_Semiring(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, a, b,
+									desc) != GrB_SUCCESS) {
+		printf("Failed adding operands, error:%s\n", GrB_error());
+		assert(false);
+	}
+
+	// Reset descriptor.
+	GrB_Descriptor_set(desc, GrB_INP0, GxB_DEFAULT);
+	GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+
+	uint child_count = AlgebraicExpression_ChildCount(exp);
+	// Expression has more than 2 operands, e.g. A+B+C...
+	for(uint i = 2; i < child_count; i++) {
+		right = exp->operation.children[i];
+
+		if(right->type == AL_OPERAND) {
+			b = right->operand.matrix;
+		} else {
+			if(right->operation.op == AL_EXP_TRANSPOSE) {
+				b = right->operation.children[0]->operand.matrix;
+				GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+			} else if(inter == GrB_NULL) {
+				// Can't use `res`, use an intermidate matrix.
+				GrB_Matrix_nrows(&nrows, res);
+				GrB_Matrix_ncols(&ncols, res);
+				GrB_Matrix_new(&inter, GrB_BOOL, nrows, ncols);
+			}
+			b = _AlgebraicExpression_Eval(right, inter);
+		}
+
+		// Perform addition.
+		if(GrB_eWiseAdd_Matrix_Semiring(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, res, b,
+										GrB_NULL) != GrB_SUCCESS) {
+			printf("Failed adding operands, error:%s\n", GrB_error());
+			assert(false);
+		}
+
+		// Reset descriptor.
+		GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
```

Comment:
Nit, but we could reset at the start of the loop and remove the call at 93.

Reply:
OK